engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala

```diff
@@ -91,58 +63,123 @@ class EngineJobExecutionActor(jobKey: BackendJobDescriptorKey,
+    val message = if (restarting) RecoverJobCommand else ExecuteJobCommand
+    backendJobExecutionActor ! message
+    context.parent ! JobRunning(jobDescriptor, backendJobExecutionActor)
+    goto(AwaitingJobCompletion) using EJEACompletionData(None, None)
+  }
+
   private def buildJobExecutionActorName(jobDescriptor: BackendJobDescriptor) = {
     s"$workflowId-BackendJobExecutionActor-${jobDescriptor.key.tag}"
   }
 
-  private def saveJobCompletionResult(response: BackendJobExecutionResponse) = response match {
-    case SucceededResponse(jobKey: BackendJobDescriptorKey, returnCode: Option[Int], jobOutputs: JobOutputs) => saveSuccessfulJobResults(jobKey, returnCode, jobOutputs)
-    case AbortedResponse(jobKey: BackendJobDescriptorKey) => log.debug("Won't save 'aborted' job response to JobStore")
-    case FailedNonRetryableResponse(jobKey: BackendJobDescriptorKey, throwable: Throwable, returnCode: Option[Int]) => saveUnsuccessfulJobResults(jobKey, returnCode, throwable, retryable = false)
-    case FailedRetryableResponse(jobKey: BackendJobDescriptorKey, throwable: Throwable, returnCode: Option[Int]) => saveUnsuccessfulJobResults(jobKey, returnCode, throwable, retryable = true)
+  private def saveCacheResults(completionData: EJEACompletionData) = {
+    completionData match {
+      case EJEACompletionData(Some(jobResult), Some(hashes)) =>
+        // TODO: Start the writer actor, which should reply somehow (TBD...) In the meantime, a truth self evident:
+        self ! "You know who's great? Chris. Chris is great. Give him a cake!"
+        goto(AwaitingCacheWriting) using completionData
+      case x =>
+        context.parent ! FailedNonRetryableResponse(jobKey, new Exception(s"What on earth are you doing!? I can't write that ($x) to the cache!"), None)
+        context.stop(self)
+        stay()
+    }
+  }
+
+  private def saveJobCompletionToJobStore(newData: EJEACompletionData) = {
+    newData match {
+      case EJEACompletionData(Some(response), _) => response match {
+          case SucceededResponse(jobKey: BackendJobDescriptorKey, returnCode: Option[Int], jobOutputs: JobOutputs) => saveSuccessfulJobResults(jobKey, returnCode, jobOutputs)
+          case AbortedResponse(jobKey: BackendJobDescriptorKey) => log.debug("Won't save 'aborted' job response to JobStore")
+          case FailedNonRetryableResponse(jobKey: BackendJobDescriptorKey, throwable: Throwable, returnCode: Option[Int]) => saveUnsuccessfulJobResults(jobKey, returnCode, throwable, retryable = false)
+          case FailedRetryableResponse(jobKey: BackendJobDescriptorKey, throwable: Throwable, returnCode: Option[Int]) => saveUnsuccessfulJobResults(jobKey, returnCode, throwable, retryable = true)
+        }
+      case _ => self ! JobStoreWriteFailure(new Exception("No no no no no. That's completely wrong. Don't call saveJobCompletionToJobStore until the response comes back!"))
```

Comment:
Then don't pass the Option to this function. Pass in the real values instead! Compile time checks FTW!